### PR TITLE
UIE-186 auth data-client decoupling

### PR DESCRIPTION
**Pull request by msilva-broad**
## Summary of changes:
Narrow example of pattern to decouple auth concerns from data-request mechanics that can be later moved to data-client shared packages.

### What
- proposed split of api fetch mechanics between ui-app (Terra UI) project and what will become a DataClient that will be moved to a xxxx-data-client package.
- implemented a narrow example of part of Runtimes ajax rework to have the V1 data client start to be built out.

### Why
- to enable and expand code-reuse of complex components that have data needs, this is a key pattern that will enable types, composition,  and package organization required for the overall strategy.

### Testing strategy
- [x] unit tests pass
- [x] added unit tests for new code
- [ ] smoke test

### CODE REVIEW

File: src/analysis/runtime-common-components.js [hunk 1]
```diff
@@ -4,8 +4,8 @@ import _ from 'lodash/fp';
 import { Fragment, useEffect, useState } from 'react';
 import { b, div, h, input, label } from 'react-hyperscript-helpers';
 import { getConvertedRuntimeStatus, usableStatuses } from 'src/analysis/utils/runtime-utils';
+import { cookiesAcceptedKey } from 'src/auth/accept-cookies';
```
Comment (msilva-broad) [inline src/analysis/runtime-common-components.js:7]: flipped dependency direction to break a circular reference unrelated to the main code changes in the PR.

**Review by jdcanas — APPROVED** [review]

File: src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { AbortOption, FetchFn } from 'src/libs/ajax/data-client-common';
+import { RawGetRuntimeItem } from 'src/libs/ajax/leonardo/models/runtime-models';
+
+export interface LeoRuntimesV1DataClient {
```
Comment (nawatts) [inline src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts:4]: Naming nits:
- This is specific to _Google_ runtimes.
- Thoughts on abbreviated vs full service names (Leo vs Leonardo)?

File: src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  details: (project: string, name: string, options?: AbortOption) => Promise<RawGetRuntimeItem>;
```
Comment (nawatts) [inline src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts]: These argument names could potentially be a little clearer:

```suggestion
  details: (googleProject: string, runtimeName: string, options?: AbortOption) => Promise<RawGetRuntimeItem>;
```

File: src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  /**
+   * fetch function that takes care of desired auth/session mechanics, api endpoint root url prefixing,
+   * baseline expected request headers, etc.
+   */
+  fetchAuthedLeo: FetchFn;
+}
+
+const runtimesPath = (project: string, name: string) => `api/google/v1/runtimes/${project}/${name}`;
+
+export const makeLeoRuntimesV1DataClient = (deps: LeoRuntimesV1DataClientDeps): LeoRuntimesV1DataClient => {
```
Comment (nawatts) [inline src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts:20]: Thoughts on generally making data clients classes?

That could possibly allow a base class to handle adding in the auth header?

Something like:
```ts
class BaseDataClient {
  constructor(authTokenSupplier) { ... }

  authenticatedRequest(url, options) {
    return fetch(url, {
      ...options,
      headers: { ...options.headers, Authorization: `Bearer ${authTokenSupplier()}` }
    });
  }
}

class LeonardoDataClient extends BaseDataClient {
  authenticatedRequest(path, options) {
    return super.authenticatedRequest(`.../${path}`, options);
  }
}
```

**Review by nawatts — APPROVED** [review]

File: src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts (hunk 4 above)
Comment (msilva-broad) [reply src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts:20]: To summarize what we chatted about on a call, I have come to embrace the elegance of (controlled) functional programming over the many pitfalls of OOP (inheritance and pervasive state-fullness being two big ones).  React hooks gives us a reasonable enough answer on how to handle state in a largely quarantined functional programming wrapper.... so I've largely embraced the FP style in the terra-ui codebase and have largely avoided classes.  Class and interface heavy solutions have their benefits, but also come with a-lot more implementation weight and rigidity.

The chainable fetch function pattern here feels like a nicely incremental solution to 95% remove the concern of auth and root config, etc. from the raw data call authoring that needs to happen in a typical data-client.... if both auth and public fetches are needed, then the depenenency list on the makeCoolDataClient function would expand to include more then one augmented fetch arg.

We can always revisit and iterate later as we learn new things.

File: src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts (hunk 3 above)
Comment (msilva-broad) [reply src/libs/ajax/leonardo/LeoRuntimesV1DataClient.ts]: good call, will fix

**Pull request merged**